### PR TITLE
[Core]Enhancement MPC  builder and solver

### PR DESCRIPTION
**Pull request by adityaghantasala**
Important points in this PR : 

- Adds process info to the steps of the master-slave constraint class (@loumalouomega  as discussed in #2942 ) and corresponding changes.

- Removing shrink_to_fit function calls on vectors at places to save memory allocation time. 

- @RiccardoRossi  I implemented a better version for doing operation K = T'*K*T and F = T'*(F - K*g). When you were here we discussed some possible bottlenecks. Now I implemented a in-place modification of the K matrix to achieve this. This version is way faster than the previous implementation.

- changes to loop for assembling the global m-s constraints.  

- Checking the size of mGlobalMasterSlaveConstraints to switch between the constraint and non constraint version (instead of number of constraints on the modelpart)

### CODE REVIEW

File: kratos/includes/master_slave_constraint.h [hunk 1]
```diff
@@ -278,33 +278,33 @@ class MasterSlaveConstraint
     /**
      * @brief This is called in the beginning of each solution step
      */
-    virtual void InitializeSolutionStep()
+    virtual void InitializeSolutionStep(ProcessInfo& rCurrentProcessInfo)
```
Comment (loumalouomega) [inline kratos/includes/master_slave_constraint.h]: :D

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h [hunk 2]
```diff
@@ -198,7 +198,7 @@ class ResidualBasedBlockBuilderAndSolverWithConstraints
         for (int k = 0; k < n_constraints; k++)
         {
             auto it = constraints_begin + k;
-            it->InitializeSolutionStep(); // Here each constraint constructs and stores its T and C matrices. Also its equation ids.
+            it->InitializeSolutionStep(rModelPart.GetProcessInfo()); // Here each constraint constructs and stores its T and C matrices. Also its equation ids.
```
Comment (loumalouomega) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h]: Better to move this outside the loop, you are accessing to the process info everytime

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h [hunk 3]
```diff
@@ -219,7 +219,7 @@ class ResidualBasedBlockBuilderAndSolverWithConstraints
         for (int k = 0; k < n_constraints; k++)
         {
             auto it = constraints_begin + k;
-            it->FinalizeSolutionStep();
+            it->FinalizeSolutionStep(rModelPart.GetProcessInfo());
```
Comment (loumalouomega) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h]: Idem

**Review by loumalouomega — DISMISSED** [review]
OK

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h [hunk 4]
```diff
@@ -602,11 +607,9 @@ class ResidualBasedBlockBuilderAndSolverWithConstraints
 
             if (constraint_is_active)
             {
-                it->EquationIdVector(slave_equation_ids, master_equation_ids, r_current_process_info);
-                //calculate constraint's T and b matrices
-                it->CalculateLocalSystem(relation_matrix, constant_vector, r_current_process_info);
                 //assemble the Constraint contribution
-                AssembleSlaves(slave_equation_ids, master_equation_ids, relation_matrix);
+                #pragma omp critical
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h:612]: why critical? this kills the parallelism...

File: kratos/utilities/helper_classes_for_constraint_builder.h [hunk 5]
```diff
@@ -605,84 +570,137 @@ class ConstraintImposer {
...
 
-        for (const auto &internal_index : rLocalIndices.internal_index_vector)
-            rTransformationMatrixLocal(internal_index, internal_index) = 1.0;
+        //Adding contribution from slave to Kmm
+        IndexType master_i = 0;
+        for (auto& master_index : mLocalIndices.processed_master_indices) {
+            IndexType master_i_other = 0;
+            for (auto& master_index_other : mLocalIndices.processed_master_indices) {
+                rLHSContribution(master_index, master_index_other) += mLocalIndices.container_master_weights[master_i] *
+                                                                        rLHSContribution(mLocalIndices.container_master_slaves[master_i], mLocalIndices.container_master_slaves[master_i_other])
```
Comment (RiccardoRossi) [inline kratos/utilities/helper_classes_for_constraint_builder.h:626]: are u accessing to a map here? take care is expensive...

File: kratos/utilities/helper_classes_for_constraint_builder.h [hunk 6]
```diff
@@ -605,84 +570,137 @@ class ConstraintImposer {
...
+            IndexType master_index = 0;
+            double master_weight = 0.0;
+            IndexType i_master = 0;
+            for (auto&  master_eq_id : master_equation_ids)
+            { // Loop over all the masters the slave has
+                master_index = std::distance(rEquationIds.begin(), std::find(rEquationIds.begin(), rEquationIds.end(), master_eq_id));
+                //master_weight = mTransformationMatrixLocal(slave_index,master_index);
+                master_weight = master_weights_vector(i_master);
+                for (auto& internal_index : mLocalIndices.internal_index_vector) {
+                    rRHSContribution(internal_index) += -rLHSContribution(internal_index, slave_index) * slave_constant;
```
Comment (RiccardoRossi) [inline kratos/utilities/helper_classes_for_constraint_builder.h]: -=

**Review by RiccardoRossi — COMMENTED** [review]
minor comments...

**Review by loumalouomega — DISMISSED** [review]
@adityaghantasala I added a Check method, could be relevant and important

**Comment by philbucher** [comment]
+1 for Check!

**Comment by adityaghantasala** [comment]
@loumalouomega  +1

File: kratos/utilities/helper_classes_for_constraint_builder.h (hunk 5 above)
Comment (adityaghantasala) [reply kratos/utilities/helper_classes_for_constraint_builder.h:626]: These are vectors @RiccardoRossi  not maps.

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h (hunk 4 above)
Comment (adityaghantasala) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h:612]: the critical is necessary for synchronized access to the map `mGlobalMasterSlaveConstraints`, which is a map of id and a unique pointer. If there is no `critical` there, there will be random segfaults.

**Review by loumalouomega — DISMISSED** [review]
Nice

**Comment by adityaghantasala** [comment]
@loumalouomega  I changed ModelPart.cpp to call the check of constraint class along with element and condition.  
Also addressed comments of @RiccardoRossi  to change to -=

File: kratos/sources/model_part.cpp [hunk 7]
```diff
@@ -1518,6 +1518,9 @@ int ModelPart::Check(ProcessInfo& rCurrentProcessInfo) const
         err = elem_iterator->Check(rCurrentProcessInfo);
     for (ConditionConstantIterator condition_iterator = ConditionsBegin(); condition_iterator != ConditionsEnd(); condition_iterator++)
         err = condition_iterator->Check(rCurrentProcessInfo);
+    for (MasterSlaveConstraintConstantIteratorType constraint_iterator = MasterSlaveConstraintsBegin(); 
+            constraint_iterator != MasterSlaveConstraintsEnd(); constraint_iterator++)
+        err = constraint_iterator->Check(rCurrentProcessInfo);
```
Comment (philbucher) [inline kratos/sources/model_part.cpp:1523]: hm won't this just return the value of the last constraint that was checked?
I guess this is also wrong in the current master

Comment (adityaghantasala) [reply kratos/sources/model_part.cpp:1523]: it does.  will suggest to leave it here and change it a different PR.

**Comment by adityaghantasala** [comment]
@RiccardoRossi  shall I proceed and merge this ? any more comments ?

**Comment by philbucher** [comment]
@adityaghantasala I just saw that you are not passing the ProcessInfo by const-reference, only by reference
Can you please change it to const-reference?
Changing it afterwards will be painful

Edit you should also pass it to `Initialize` and `Finalize`

See discussions in #1897 and #2993 to pass ProcessInfo by const-ref to Conditions and Elements

**Comment by philbucher** [comment]
Also `GetDofList` and `EquationIdVector` should be const functions, see #732

**Comment by adityaghantasala** [comment]
Done !

**Review by loumalouomega — DISMISSED** [review]

**Comment by philbucher** [comment]
can you please pass the processinfo also to Initialize and Finalize? => #2817

**Comment by adityaghantasala** [comment]
sorry missed it .. added it ..

**Review by philbucher — DISMISSED** [review]
Thanks!

One question: Aren't you calling initialize & finalize somwhere? if so, then also there the ProcessInfo should be passed

**Comment by adityaghantasala** [comment]
good question .. I am not calling them ... but they should be called in scheme's initialize function (elements and conditions are there). we can change that in the next pr.

**Comment by philbucher** [comment]
ok
All good from my side then

**Comment by loumalouomega** [comment]
I was speaking with @RiccardoRossi , and not for this PR but for a future one. It would be interesting to integrate inside the MasterSlavConstraint its own EquationIdVector method

**Comment by adityaghantasala** [comment]
@loumalouomega  I cannot see the application now. But yes .. it is possible to do it like that.

**Comment by adityaghantasala** [comment]
I am proceeding with merging this. the suggested changes to modelpart.cpp will be done in a different PR.

**Comment by loumalouomega** [comment]
The application would be to add equation id of dof not added directly by other conditions or elements

**Comment by adityaghantasala** [comment]
Then it is also important that it reflects in GetDofsVector function.

**Comment by loumalouomega** [comment]
Of course, I meant the couple

**Comment by msandre** [comment]
@adityaghantasala can you wait until start of next week so I can take a look?

File: kratos/includes/master_slave_constraint.h [hunk 8]
```diff
@@ -359,6 +359,25 @@ class MasterSlaveConstraint
...
+     * @details It is designed to be called only once (or anyway, not often) typically at the beginning
+     * of the calculations, so to verify that nothing is missing from the input or that no common error is found.
+     * @param rCurrentProcessInfo
+     * @note This method is: MANDATORY
+     */
+    virtual int Check(const ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_TRY
+
+        KRATOS_ERROR_IF( this->Id() < 1 ) << "MasterSlaveConstraint found with Id " << this->Id() << std::endl;
```
Comment (RiccardoRossi) [inline kratos/includes/master_slave_constraint.h:374]: this only throws if the id is zero no?

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h [hunk 9]
```diff
@@ -213,13 +218,18 @@ class ResidualBasedBlockBuilderAndSolverWithConstraints
     {
         KRATOS_TRY
         BaseType::FinalizeSolutionStep(rModelPart, A, Dx, b);
+	
+	// Getting process info
+	ProcessInfo& r_process_info = rModelPart.GetProcessInfo();
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h]: const ProcessInfo& ?

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h (hunk 4 above)
Comment (RiccardoRossi) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h:612]: yes i assumed something of the sort :-). The point is: this critical is there only in the construction phase or any time you do build?

**Review by RiccardoRossi — COMMENTED** [review]
in general i am fine with this. just few minor comments

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h [hunk 10]
```diff
@@ -192,13 +192,18 @@ class ResidualBasedBlockBuilderAndSolverWithConstraints
         KRATOS_TRY
 
         BaseType::InitializeSolutionStep(rModelPart, A, Dx, b);
+	
+	// Getting process info
+	ProcessInfo& r_process_info = rModelPart.GetProcessInfo();
```
Comment (msandre) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h]: const

File: kratos/utilities/helper_classes_for_constraint_builder.h [hunk 11]
```diff
@@ -605,84 +572,137 @@ class ConstraintImposer {
...
+     *          Meaning that there is no memory duplication and no explicit matrix and matrix or matrix vector multiplication.
+     *          Individual entries of K and F are modified to achieve the result.
+     * @param   rLHSContribution The lhs matrix of the container
+     * @param   rRHSContribution The rhs vector of the container
+     * @param   rEquationIds the list of equation ids (extended with the masters).
      */
-    void CalculateLocalTransformationMatrix(LocalIndicesType& rLocalIndices,
-                                            MatrixType& rTransformationMatrixLocal,
-                                            EquationIdVectorType& rEquationIds)
+    void ModifyLHS(MatrixType &rLHSContribution, VectorType& rRHSContribution, EquationIdVectorType &rEquationIds)
```
Comment (msandre) [inline kratos/utilities/helper_classes_for_constraint_builder.h]: Could you make the function name more descriptive?

**Review by msandre — COMMENTED** [review]
Some minor comments from my side. When things have settled, it would be good to add a wiki describing the overall class design and how it interfaces with the other core classes.

**Comment by adityaghantasala** [comment]
@RiccardoRossi  yes the critical is only in construction phase (initialization) and **does not** happen every time we build (of course then the constraint relations should stay constant)

File: kratos/includes/master_slave_constraint.h (hunk 8 above)
Comment (adityaghantasala) [reply kratos/includes/master_slave_constraint.h:374]: is 0 and less than zero or ?

**Comment by adityaghantasala** [comment]
@msandre  I addressed your comments also

**Comment by adityaghantasala** [comment]
sorry for noise .. I committed some local files.

**Review by RiccardoRossi — APPROVED** [review]

**Comment by loumalouomega** [comment]
@adityaghantasala you forgot to remove the branch

**Pull request merged**